beehive-lab/TornadoVM
# Fix dataflow analysis to retrieve data accesses

#### Description

This PR fixes the following:
- dataflow analysis to retrieve data accesses when not fully inlining.
- parameter load order in the callee when some of the parameters of the callee are not used.
- NPE in the scheduler when performing task migration with a CPU device.
- Methods/functions invoked from different places in the call graph do not get compiled twice.

I also included some cleanup of unused methods (`OCLCompilationResultsBuilder` and `OCLCompiler`)  and classes.

#### Problem description

In the Sketcher, Tornado runs a `TornadoDataflowAnalysis` phase that looks at the dataflow and assigns an `Access` for each parameter. Currently, we only analyze the method passed to the task, but not the non-inlined methods in the call graph. This can cause wrong accesses to be assigned to the parameters. Accesses of tasks are then used to generate nodes in the `TornadoGraph`.

With this PR, we merge the accesses of the caller with the accesses of the callees, obtaining the right accesses for the method.

Additionally, with the call graph below, method D could get compiled twice. This PR fixes this in the `OCL/PTXCompiler` and tests it in `TestMultipleFunctions::testNoDoubleCompilation` 

```
A → B → D
  ↘ C ↗
```
#### Backend/s tested


#### OS tested


#### Did you check on FPGAs?

If possible, check your changes on FPGAs.


#### How to test the new patch?

Run all the unit tests. 
`TestMultipleFunctions` should check for the task accesses and parameter order in non-inlined callees.
